Integrate electronics

CJBuchel opened the pull request:
- Main outline for electronics
- Separate compiled modes for microcontrollers (RAM/BAM/STM/SGM)
- libs created for Buttons/Displays/Beam Breaks/E Stops
- Element classes made for auto-scoring (Not applied)

JaciBrunning commented on electronics/platformio.ini:
```diff
@@ -12,19 +12,11 @@
-
-; debug_tool = stlink
-
-; Use our fixed board definition for debug and upload, since the stock one that comes with
```
Please bring back the comments, they were there for a reason

CJBuchel replied:
oh, whoops. I deleted that by accident when I was trying to fix something by adding flags

JaciBrunning commented on electronics/src/Config.h:
```diff
@@ -0,0 +1,103 @@
+ * 
+ */
+
+// #define RAM // Red Alliance controller mode
```
"RAM" could be misconstrued as Random Access Memory instead of Red Alliance Mode - maybe another name is a better idea

CJBuchel replied:
I'll think of a new name

JaciBrunning commented on electronics/src/Config.h:
```diff
@@ -0,0 +1,103 @@
+#define MODE 3
+#endif
+
+#define global static const // Constant experssion accessable to all (cannot be modified)
```
Giving this a define is an odd decision - it hides the meaning that "static const" gives that C++ programmers are used to.

CJBuchel replied:
Yea, not very needed. I'll switch it back, this was for something fancy when global was a shortcut for a very long string of C++ keywords, but it ended up being only the two.

JaciBrunning commented on electronics/src/Config.h:
```diff
@@ -0,0 +1,103 @@
+ * Config for Power port
+ */
+#ifdef RAM
+#define IP "10.0.100.4"
```
IP for all devices should be given by DHCP, not static. Devices will report to the FMS and advertise their "role" in the first message

JaciBrunning replied:
Let's also look at the possibility of setting the "mode"/"role" at runtime - we can store it in flash like in my example which would avoid needing to reflash controllers every time we setup.

CJBuchel replied:
Ip and port stuff is mostly just a placeholder, we should probably go through what we actually want to do with communications. 

E.g TCP or UDP, standard buffer size? maybe 256? etc...
I'm not completely sure what networking you're imagining, so soon we need to go over some kind of network map, and how we want everything being done

JaciBrunning commented on electronics/src/Handles.h:
```diff
@@ -0,0 +1,75 @@
+ * Handler for main controllers,
+ * (insert interrupt code)
+ */
+#define HandleController(Controller) \
```
Let's avoid using threads on the microcontrollers. There's no real reason for the overhead of threads to be introduced - for this purpose let's just keep it simple: Read -> Process -> Write

CJBuchel replied:
There shouldn't be any threads in there. That user button was deleted, but it seems to have found it's way back in. That will be removed.

CJBuchel commented:
FIxed up code to comply with changes.

For the moment I've also added an "AUTO_CONFIG" mode in config.h, which we might use if we get time for it, applying your idea of having the Nucleo get it'd role by runtime. Possibly over the network from JMS setting it as either red alliance or blue alliance etc...
